### PR TITLE
Analyzer file reference

### PR DESCRIPTION
**Pull request by tmat**
Enables tests on Core CLR.
Adds more tests.

Checks that the path passed to the constructor is actually a full path. If it isn't the behavior of the reference is non-deterministic as it depends on the process-wide current directory.
Reimplements `Equals` and `GetHashCode` so that they do not load the assembly. This is critical to avoiding loading unnecessary assemblies to the IDE.

### CODE REVIEW

**Comment by tmat** [comment]
@jasonmalinowski @mavasani @dotnet/roslyn-compiler PTAL

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+using System;
+using System.Collections.Immutable;
+using System.Reflection;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Roslyn.Test.Utilities.Desktop;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.UnitTests
```
Comment (tmat) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs:18]: >namespace M [](start = 0, length = 11)

These tests are just moved from the original file. No changes made.

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 Microsoft.CodeAnalysis.CommandLineSourceFile.CommandLineSourceFile(string path, bool isScript, bool isInputRedirected) -> void
 Microsoft.CodeAnalysis.CommandLineSourceFile.IsInputRedirected.get -> bool
+Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference.Equals(Microsoft.CodeAnalysis.Diagnostics.AnalyzerFileReference other) -> bool
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:3]: Did we really want to make this public? If you can call the other one with a AnalyzerFileReference and get the same behavior, is it really needed?

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 3]
```diff
@@ -53,19 +55,58 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
...
-                assemblyLoader.AddDependencyLocation(fullPath);
+                return true;
             }
+
+            return other is object &&
+                ReferenceEquals(_assemblyLoader, other._assemblyLoader) &&
+                FullPath == other.FullPath;
         }
 
+        // legacy, for backwards compat:
```
Comment (jasonmalinowski) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:86]: Is this worth Obsoleting because it's behavior is, well, funky?

File: src/Compilers/Core/Portable/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (tmat) [reply src/Compilers/Core/Portable/PublicAPI.Unshipped.txt:3]: I guess it's not needed, just a common pattern for Equality/GHC.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:86]: Maybe. I'll let compiler team to decide.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 4]
```diff
@@ -53,19 +55,58 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
         /// <param name="assemblyLoader">Loader for obtaining the <see cref="Assembly"/> from the <paramref name="fullPath"/></param>
         public AnalyzerFileReference(string fullPath, IAnalyzerAssemblyLoader assemblyLoader)
         {
-            _fullPath = fullPath ?? throw new ArgumentNullException(nameof(fullPath));
+            CompilerPathUtilities.RequireAbsolutePath(fullPath, nameof(fullPath));
```
Comment (mavasani) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:58]: Can this break existing hosts that use non-absolute path here, but have baked in special knowledge in the provided assembly loader to correctly load the assembly?

Comment (tmat) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:58]: Yes. But that scenario does not work anyways because we don't use assembly loader for reading the Id/DisplayName.

Comment (mavasani) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:58]: Sounds good.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 5]
```diff
@@ -53,19 +55,58 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
...
+        {
+            if (ReferenceEquals(this, other))
             {
-                assemblyLoader.AddDependencyLocation(fullPath);
+                return true;
             }
+
+            return other is object &&
+                ReferenceEquals(_assemblyLoader, other._assemblyLoader) &&
+                FullPath == other.FullPath;
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:83]: I find our path story on case sensitive or insensitive hard to follow. Would've expected this to be case insensitive but you're just going with the existing behavior here 🤷

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs (hunk 3 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:86]: @333fred

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        public Exception LoadAnalyzer(string analyzerPath)
+        {
+            _analyzerLoadException = null;
+            var analyzerRef = new AnalyzerFileReference(analyzerPath, FromFileLoader.Instance);
+            analyzerRef.AnalyzerLoadFailed += (s, e) => _analyzerLoadException = e.Exception;
+            var builder = ImmutableArray.CreateBuilder<DiagnosticAnalyzer>();
+            analyzerRef.AddAnalyzers(builder, LanguageNames.CSharp);
+            return _analyzerLoadException;
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs:36]: Why is `_analyzerLoadException` being stored as a field here? It's only ever used in this method

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs (hunk 5 above)
Comment (tmat) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:83]: Yes, keeping the current behavior. I am open to discussing changing it but that'd be a separate change.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs (hunk 6 above)
Comment (tmat) [reply src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceAppDomainTests.cs:36]: IDK. Seems like it could be a local. Fixing.

File: src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs [hunk 7]
```diff
@@ -36,48 +33,68 @@ public Assembly LoadFromPath(string fullPath)
...
+            // different loader:
+            Assert.False(refA.Equals(new AnalyzerFileReference(path1, loader2)));
+
+            // legacy overload:
+            Assert.True(refA.Equals((AnalyzerReference)refA));
+            Assert.False(refA.Equals((AnalyzerReference?)null));
+            Assert.True(refA!.Equals((AnalyzerReference)refB));
+            Assert.True(refA.Equals(new TestAnalyzerReference(path1)));
+            Assert.False(refA.Equals(new TestAnalyzerReference(path2)));
         }
```
Comment (cston) [inline src/Compilers/Core/CodeAnalysisTest/Analyzers/AnalyzerFileReferenceTests.cs:98]: >} [](start = 8, length = 1)

Consider adding `Assert.Equal(refA.GetHashCode(), refB.GetHashCode());`.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs [hunk 8]
```diff
@@ -53,19 +55,58 @@ public sealed class AnalyzerFileReference : AnalyzerReference, IEquatable<Analyz
...
+            {
+                return false;
+            }
+
+            if (other is AnalyzerFileReference fileReference)
+            {
+                return Equals(fileReference);
+            }
+
+            return FullPath == other.FullPath;
```
Comment (cston) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:104]: >FullPath == other.FullPath [](start = 19, length = 26)

When is this `true` for distinct `AnalyzerReference` types?

Comment (tmat) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:104]: This is for backward compat. We shouldn't really be comparing different type of references, but we did implement it this way so need to keep it.
Note that anyone can subclass AnalyzerReference and then compare AnalyzerFileReference to it.

File: src/Compilers/Core/Portable/FileSystem/PathUtilities.cs [hunk 9]
```diff
@@ -400,29 +400,29 @@ private static bool IsDriveRootedAbsolutePath(string path)
...
         /// Combine two paths, the first of which may be absolute.
         /// </summary>
-        /// <param name="rootOpt">First path: absolute, relative, or null.</param>
+        /// <param name="root">First path: absolute, relative, or null.</param>
         /// <param name="relativePath">Second path: relative and non-null.</param>
-        /// <returns>null, if <paramref name="rootOpt"/> is null; a combined path, otherwise.</returns>
+        /// <returns>null, if <paramref name="root"/> is null; a combined path, otherwise.</returns>
         /// <seealso cref="CombineAbsoluteAndRelativePaths"/>
-        public static string? CombinePossiblyRelativeAndRelativePaths(string? rootOpt, string? relativePath)
+        public static string? CombinePossiblyRelativeAndRelativePaths(string? root, string? relativePath)
```
Comment (cston) [inline src/Compilers/Core/Portable/FileSystem/PathUtilities.cs:407]: >string? root, string? relativePath [](start = 70, length = 34)

It looks like the only caller is `CombineAbsoluteAndRelativePaths`. If that's the case, consider changing the parameters to `string`.

Comment (tmat) [reply src/Compilers/Core/Portable/FileSystem/PathUtilities.cs:407]: There are other callers in the IDE. I think it's ok to accept null root since we need to check for empty string anyways...

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs (hunk 3 above)
Comment (tmat) [reply src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerFileReference.cs:86]: I'll follow up.

**Comment by sharwell** [comment]
> Require absolute path

This change broke AnalyzerRunner.
